File: 2025-09-25 - Deques/Exemplo_Python/ListaPrioridadePQ.java
// ------------------------------------------------------------------------
// ListaPrioridadePQ.java
// Implementação da Lista de Prioridade usando PriorityQueue
// ------------------------------------------------------------------------
import java.util.PriorityQueue;
import java.util.Iterator;

public class ListaPrioridadePQ implements ListaPrioridadeInterface {

   private PriorityQueue<Node> pq;

   // ---------------------------------------------------------------------
   // Construtor
   // ---------------------------------------------------------------------
   public ListaPrioridadePQ() {
      this.pq = new PriorityQueue<>();
   }

   // ---------------------------------------------------------------------
   // Inserir elemento
   // ---------------------------------------------------------------------
   @Override
   public boolean inserir(Node novoNode) {
      if (novoNode == null) return false;
      // Evitar duplicidade
      if (buscar(novoNode.getCodigoDisciplina()) != null) return false;
      return pq.offer(novoNode);
   }

   // ---------------------------------------------------------------------
   // Remover por código da disciplina
   // ---------------------------------------------------------------------
   @Override
   public boolean remover(String codigoDisciplina) {
      Node node = buscar(codigoDisciplina);
      if (node != null) {
         return pq.remove(node);
      }
      return false;
   }

   // ---------------------------------------------------------------------
   // Buscar por código da disciplina
   // ---------------------------------------------------------------------
   @Override
   public Node buscar(String codigoDisciplina) {
      for (Node node : pq) {
         if (node.getCodigoDisciplina().equals(codigoDisciplina)) return node;
      }
      return null;
   }

   // ---------------------------------------------------------------------
   // Atualizar disciplina existente
   // ---------------------------------------------------------------------
   @Override
   public boolean atualizar(Node nodeAtualizado) {
      if (nodeAtualizado == null) return false;
      if (remover(nodeAtualizado.getCodigoDisciplina())) {
         return inserir(nodeAtualizado);
      }
      return false;
   }

   // ---------------------------------------------------------------------
   // Imprimir lista completa
   // ---------------------------------------------------------------------
   @Override
   public void imprimirLista() {
      // Como PriorityQueue não garante ordem na iteração,
      // copiamos para um array e ordenamos para exibir
      PriorityQueue<Node> copia = new PriorityQueue<>(pq);
      System.out.println("Lista de Prioridade (ordenada por semestre):");
      while (!copia.isEmpty()) {
         System.out.println(copia.poll());
      }
   }

   // ---------------------------------------------------------------------
   // Retornar tamanho
   // ---------------------------------------------------------------------
   @Override
   public int tamanho() {
      return pq.size();
   }
}

File: 2025-09-25 - ListaPrioridadePQ/Exemplo_C#/ListaPrioridadePQ.cs
using System;
using System.Collections.Generic;

// ------------------------------------------------------------------------
// listaPrioridadePQ.cs
// Implementação da lista de prioridade usando PriorityQueue
// Todos os métodos e variáveis em camelCase
// ------------------------------------------------------------------------
public class ListaPrioridadePQ : ListaPrioridadeInterface
{
    private PriorityQueue<Node, int> pq;

    // Construtor
    public ListaPrioridadePQ()
    {
        pq = new PriorityQueue<Node, int>();
    }

    // ---------------------------------------------------------------------
    // Inserir elemento mantendo a prioridade (menor semestre = maior prioridade)
    // ---------------------------------------------------------------------
    public bool inserir(Node novoNode)
    {
        if (novoNode == null) return false;
        if (buscar(novoNode.codigoDisciplinaProp) != null) return false; // evita duplicidade
        pq.Enqueue(novoNode, novoNode.semestreProp);
        return true;
    }

    // ---------------------------------------------------------------------
    // Remover elemento pelo código da disciplina
    // ---------------------------------------------------------------------
    public bool remover(string codigoDisciplina)
    {
        List<(Node, int)> temp = new List<(Node, int)>();
        bool encontrado = false;

        while (pq.Count > 0)
        {
            pq.TryDequeue(out Node node, out int prioridade);
            if (node.codigoDisciplinaProp == codigoDisciplina)
            {
                encontrado = true;
                continue; // remove este elemento
            }
            temp.Add((node, prioridade));
        }

        foreach (var item in temp)
        {
            pq.Enqueue(item.Item1, item.Item2);
        }

        return encontrado;
    }

    // ---------------------------------------------------------------------
    // Buscar elemento pelo código da disciplina
    // ---------------------------------------------------------------------
    public Node buscar(string codigoDisciplina)
    {
        List<(Node, int)> temp = new List<(Node, int)>();
        Node encontrado = null;

        while (pq.Count > 0)
        {
            pq.TryDequeue(out Node node, out int prioridade);
            if (node.codigoDisciplinaProp == codigoDisciplina)
            {
                encontrado = node;
            }
            temp.Add((node, prioridade));
        }

        foreach (var item in temp)
        {
            pq.Enqueue(item.Item1, item.Item2);
        }

        return encontrado;
    }

    // ---------------------------------------------------------------------
    // Atualizar disciplina existente
    // ---------------------------------------------------------------------
    public bool atualizar(Node nodeAtualizado)
    {
        if (nodeAtualizado == null) return false;
        if (remover(nodeAtualizado.codigoDisciplinaProp))
        {
            return inserir(nodeAtualizado);
        }
        return false;
    }

    // ---------------------------------------------------------------------
    // Imprimir lista completa (ordenada por semestre)
    // ---------------------------------------------------------------------
    public void imprimirLista()
    {
        List<(Node, int)> temp = new List<(Node, int)>();

        Console.WriteLine("Lista de Prioridade (ordenada por semestre):");
        while (pq.Count > 0)
        {
            pq.TryDequeue(out Node node, out int prioridade);
            Console.WriteLine(node.toStringNode());
            temp.Add((node, prioridade));
        }

        foreach (var item in temp)
        {
            pq.Enqueue(item.Item1, item.Item2);
        }
    }

    // ---------------------------------------------------------------------
    // Retornar tamanho da lista
    // ---------------------------------------------------------------------
    public int tamanho()
    {
        return pq.Count;
    }
}
